utils/src/offset/zero_copy.rs

```diff
@@ -0,0 +1,283 @@
+use std::mem;
+
+/// Casts a part of provided `bytes` buffer with the given `offset` to a
+/// mutable pointer to `T`.
+///
+/// Should be used for single values.
+///
+/// # Safety
+///
+/// This is higly unsafe. This function doesn't ensure alignment and
+/// correctness of provided buffer. The responsibility of such checks is on
+/// the caller.
+pub unsafe fn read_ptr_at<T>(bytes: &[u8], offset: &mut usize) -> *mut T {
+    let size = mem::size_of::<T>();
+    let ptr = bytes[*offset..*offset + size].as_ptr() as *mut T;
+    *offset += size;
+    ptr
+}
+
+/// Casts a part of provided `bytes` buffer with the given `offset` to a
+/// mutable pointer to `T`.
+///
+/// Should be used for array-type sequences.
+///
+/// # Safety
+///
+/// This is higly unsafe. This function doesn't ensure alignment and
+/// correctness of provided buffer. The responsibility of such checks is on
+/// the caller.
+pub unsafe fn read_array_like_ptr_at<T>(bytes: &[u8], offset: &mut usize, len: usize) -> *mut T {
+    let size = mem::size_of::<T>() * len;
+    let ptr = bytes[*offset..*offset + size].as_ptr() as *mut T;
+    *offset += size;
+    ptr
+}
+
+/// Writes provided `data` into provided `bytes` buffer with the given
+/// `offset`.
+pub fn write_at<T>(bytes: &mut [u8], data: &[u8], offset: &mut usize) {
+    let size = mem::size_of::<T>();
+    bytes[*offset..*offset + size].copy_from_slice(data);
+    *offset += size;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use bytemuck::{Pod, Zeroable};
+    use memoffset::offset_of;
+
+    #[test]
+    fn test_read_ptr_at() {
+        #[derive(Clone, Copy, Pod, Zeroable)]
+        #[repr(C)]
+        struct TestStruct {
+            a: isize,
+            b: usize,
+            c: i64,
+            d: u64,
+            e: i32,
+            f: u32,
+            g: i16,
+            _padding_1: [u8; 2],
+            h: u16,
+            _padding_2: [u8; 2],
+            i: i8,
+            _padding_3: [i8; 3],
+            j: u8,
+            _padding_4: [i8; 3],
+        }
+
+        let mut buf = vec![0_u8; mem::size_of::<TestStruct>()];
+        let s = buf.as_mut_ptr() as *mut TestStruct;
+
+        unsafe {
+            (*s).a = -9223372036854771632;
+            (*s).b = 9223372036854771632;
+            (*s).c = -9223372036854771632;
+            (*s).d = 9223372036854771632;
+            (*s).e = -2147483623;
+            (*s).f = 2147483623;
+            (*s).g = -32721;
+            (*s).h = 32721;
+            (*s).i = -127;
+            (*s).j = 127;
```

Comment:
More magic numbers, other wise looks good feel free to merge after rebase and fix :)